cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java

```diff
@@ -0,0 +1,361 @@
+  protected void triggerShutdown() {
+    stopped = true;
+    if (runThread != null) {
+      runThread.interrupt();
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.debug("Shutting down log processor pipeline for {}", name);
+
+    try {
+      context.stop();
+      // Persist the checkpoints. It can only be done after successfully stopping the appenders.
+      // Since persistCheckpoint never throw, putting it inside try is ok.
+      persistCheckpoints();
+    } catch (Exception e) {
+      // Just log, not to fail the shutdown
+      LOG.warn("Exception raised when stopping pipeline {}", name, e);
+    }
+
+    LOG.info("Log processor pipeline for {} stopped with latest checkpoint {}", name, checkpoints);
+  }
+
+  @Override
+  protected String getServiceName() {
+    return "LogPipeline-" + name;
+  }
+
+  /**
+   * Reads events from incomingEventQueue and sends them to event processor for further processing.
+   */
+  private boolean processEvents(BlockingQueue<LogBufferEvent> incomingEventQueue) {
+    ProcessedEventMetadata<LogBufferFileOffset> metadata
+      = eventQueueProcessor.process(0, new LogFileOffsetTransformIterator(incomingEventQueue));
+
+    // none of the events were processed.
+    if (metadata.getTotalEventsProcessed() <= 0) {
+      return false;
+    }
+
+    unSyncedEvents += metadata.getTotalEventsProcessed();
+    // events were processed, so update the checkpoints
+    Checkpoint<LogBufferFileOffset> checkpoint = metadata.getCheckpoints().get(0);
+    MutableLogBufferCheckpoint mutableCheckpoint = checkpoints.get(0);
+    MutableLogBufferFileOffset offset = mutableCheckpoint.getOffset();
+    offset.setFileId(checkpoint.getOffset().getFileId());
+    offset.setFilePos(checkpoint.getOffset().getFilePos());
+    mutableCheckpoint.setMaxEventTs(checkpoint.getMaxEventTime());
+
+    return true;
+  }
+
+  /**
+   * This method is called from {@link ConcurrentLogBufferWriter} to push log events.
+   *
+   * @param events log events to be processed
+   */
+  public void processLogEvents(Iterator<LogBufferEvent> events) {
+    // Don't accept any log events if the pipeline is not running
+    if (!isRunning()) {
+      return;
+    }
+
+    while (events.hasNext()) {
+      try {
+        // This call will block ConcurrentLogBufferWriter thread until the queue has free space.
+        incomingEventQueue.put(events.next());
+      } catch (InterruptedException e) {
+        // Just ignore the exception and reset the flag
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Persists the checkpoints.
+   */
+  private void persistCheckpoints() {
+    try {
+      checkpointManager.saveCheckpoints(checkpoints);
+      LOG.debug("Checkpoint persisted for {} with {}", name, checkpoints);
+    } catch (Exception e) {
+      // Just log as it is non-fatal if failed to save checkpoints
+      OUTAGE_LOG.warn("Failed to persist checkpoint for pipeline {}.", name, e);
```

Comment:
Are we sure we want to log and continue if the checkpoint can't be saved? What types of exceptions can be thrown when saving checkpoints? Do we want to retry?

Reply:
If the checkpointing failed when called from `trySyncAndPersistCheckpoints` method, it will be retried again by run method in next iteration. However, if the its called from shutdown method then it won't be retried.

Reply:
But is it correct to continue on in that scenario? For example, if the checkpoint store is down for some time, that means it will keep syncing log events even though the checkpoint is far behind. What happens if the log service then dies? Will it rewrite all those logs again?

Reply:
Log saver makes sure there is no loss of events in exception cases. However, there may be log duplication. Added comments